### ISSUE TITLE
IBM OS/2 1.3 Emulation issue

### ISSUE THREAD

**Issue by DropboxKenshiro** [86Box/86Box#383]
So, I tried to install IBM OS/2 1.3 just for fun, but when I run the first installation floppy, it returns a critical error (output is on screenshot). That makes installing the whole OS impossible. One of Discord users suggested that the reason is unexpected division by zero. I checked media and it's correct. I also inlcude machine's config. Of course I don't give the OS itself, you can grab it from the net if you want.

Error output:
![20190505_113415](https://user-images.githubusercontent.com/30048944/57191794-bd31a380-6f29-11e9-9034-836c58e6c702.png)

Machine config: https://pastebin.com/1LM81nc2

Because it's an installation error, no HDD image included.

**Comment by ppgrainbow** [comment]
If you're trying to install OS/2 1.3 on a hard disk that is bigger than 1,024 cylinders, 16 heads or 63 sectors per track, you're going to get this error. This is the case even if you try to install it on a machine that has LBA support.

If you want to install OS/2 1.3, you will need to run it on a hard disk that is 1,024 cylinders or smaller or use a SCSI hard disk.

**Comment by daviunic** [comment]
This error doesn't happen on Award 486, but it does on AMI Win486. Same hard disk size.

**Comment by ppgrainbow** [comment]
> 
> 
> This error doesn't happen on Award 486, but it does on AMI Win486. Same hard disk size.

I'm wondering if it's also happening on the board that uses the ALi-1429 chipset as the ALi-1429 chipset emulation is not 100% complete.

**Comment by waltje** [comment]
Most likely depends on how the BIOS handles it, and how it is translated once
the OS/2 kernel takes over.

--fred

--------------------------------------------
On Mon, 5/6/19, ppgrainbow <notifications@github.com> wrote:

 Subject: Re: [86Box/86Box] IBM OS/2 1.3 Emulation issue (#383)
 To: "86Box/86Box" <86Box@noreply.github.com>
 Cc: "Subscribed" <subscribed@noreply.github.com>
 Date: Monday, May 6, 2019, 4:46 PM
 
 
 This error doesn't happen on Award 486, but it does
 on AMI Win486. Same hard disk size.
 
 I'm wondering if it's also happening on the board
 that uses the ALi-1429 chipset as the ALi-1429 chipset
 emulation is not 100% complete.
 
 —
 You are receiving this because you are subscribed to this
 thread.
 Reply to this email directly, view
 it on GitHub, or mute
 the thread.

**Comment by daviunic** [comment]
It doesn't have anything to do with the hard disk. Same error on AMI 486 boards with or without a hard drive. Other hardware has no effect on it either it seems.

**Comment by ppgrainbow** [comment]
I'm gonna try the OS/2 1.3 installation disks under PCem v15 to see if this can be reproduced.

**Comment by ppgrainbow** [comment]
I have tested IBM OS/2 1.30.1 and the OS crashed with a TRAP 0000 referencing SCREEN$.

IBM OS/2 1.30.2 does work. What I had to do was to boot into MS-DOS first, perform a FDISK and format a hard disk image (up to 504 MB on IDE and up to 1 GB on SCSI). After that, I then proceed to install OS/2 1.30.2 on to the OPTi-495SX-based 486DX2/66 as usual. :)

Taken from the Installing OS/2 1.x in a VirtualBox VM from the OS/2 Museum. Unfortunately, I can't access the OS/2 Museum webpage right now because the server is down:

_The crash is a division by zero (trap 0) on a DIV CX instruction where the CX register contains zero. The way around this problem is patching OS/2 and getting rid of the bad loop. The original code sequence looks like this:

```
    MOV  DX, 3
L1:
    XOR  CX, CX
L2:
    LOOP L2        ; loop 65,536 times
    DEC  DX        ; run the loop 3 times
    JNE  L1
    CLI
    ...            ; read current tick
    STI
    SUB  CX, BX
    MOV  AX, 1F4h  ; 500
    MOV  BX, 0C8h  ; 200
    MUL  BX        ; 500 * 200 = 100,000
    DIV  CX        ; traps if CX = 0
    SHR  CX, 1

```
This obviously became a problem as far back as 1991, because OS/2 1.30.1 contains slightly different code:

```
    MOV  BX, 0C8h ; 200
    MUL  BX
    CMP  CX, 0
    JNE  L3
    MOV  CX, 1Fh  ; 31
L3:
    DIV  CX
```

In other words, if CX is zero, it is forced to a non-zero value to avoid the division by zero trap.

As it turns out, the MOV/MOV/MUL/DIV instruction sequence is quite unique and occurs only in the BASEDD0x.SYS drivers. To avoid the crash, it is sufficient to remove the DIV instruction, replacing it by NOPs._

To summarise, if you want to install IBM OS/2 1.3 on a faster PC, you will have to use a HEX editor and replace all occurrences of:

`B8 F4 01 BB C8 00 F7 E3 F7 F1
`
With the following
`
B8 F4 01 BB C8 00 F7 E3 90 90`

I don't know lines in this code would be in, but it would be helpful.

Anyways, should we mark this as WONTFIX in IBM OS/2 1.3 since this bug was corrected in IBM OS/2 1.30.2?

Thanks. :)

**Comment by daviunic** [comment]
Apparently an OS bug, not an emulator issue => wontfix.
